### PR TITLE
Feat/current campaign group

### PR DESCRIPTION
**Pull request by jeffersonrabb**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Together with https://github.com/Automattic/newspack-popups/pull/364 adds the concept of the active campaign group. Only campaigns in the active group will be shown to end users. If there is no active group, no campaigns will be shown. 

**Note that this is a "breaking" change since it means legacy campaigns will stop working as soon as these changes take effect. I'm working on  few possible migration strategies to convert legacy campaign configurations to the new way, including this [CLI-based approach](https://github.com/Automattic/newspack-popups/tree/feat/cli-converter). We'll need to be careful not to deploy these branches until a migration solution is settled on.**

Screen Shot 2021-01-03 at 3 58 22 PM

Dependent PRs:
~https://github.com/Automattic/newspack-plugin/pull/771~
~https://github.com/Automattic/newspack-plugin/pull/769~
~https://github.com/Automattic/newspack-plugin/pull/768~
~https://github.com/Automattic/newspack-plugin/pull/767~

### How to test the changes in this Pull Request:

Switch `newspack-popups` to `feat/current-campaign-group`.

1. Make sure some Campaign Groups have been created. For now, this can only be done in the sidebar of the edit screen for a campaign. 
2. Navigate to Newspack->Campaigns->Campaigns. In the Campaign Groups `SelectControl` select a group. Observe an "Activate" button appears next to it.
3. Click to activate. Observe the label in the `SelectControl` now has "(Active)" after it. Observe there is a "Deactivate" button next to it.
4. Refresh the page, observe the Campaign Groups UI defaults to the active group. Observe this is also true when switching from another tab back to Campaigns.
5. Switch back to "All Campaigns." Assign various Campaign Groups to different campaigns. Note: the work adding Campaign Group UI to the popover is in https://github.com/Automattic/newspack-plugin/pull/767, which is merged into this branch.
6. Experiment with activating various Campaign Groups (including none) and verify in an incognito window that end users only see campaigns in the active group.
7. Verify that any individual campaign can be previewed, whether it is in the active group or not. Verify from both the Campaigns Wizard (click three dots on right->preview) and from within the editor sidebar.
8. Verify that other groups can be previewed successfully, whether active or not. Try this from both the Campaigns and Preview tabs.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: assets/wizards/popups/views/popup-group/index.js [hunk 1]
```diff
@@ -59,22 +61,36 @@ const descriptionForPopup = (
...
 
 	useEffect( () => {
 		apiFetch( {
 			path: '/wp/v2/newspack_popups_taxonomy?_fields=id,name',
 		} ).then( terms => setCampaignGroups( terms ) );
 	}, [] );
 
+	useEffect( () => {
+		if ( -1 === campaignGroup && activeCampaignGroup > 0 ) {
+			setCampaignGroup( activeCampaignGroup );
```
Comment (dkoo) [inline assets/wizards/popups/views/popup-group/index.js:90]: I think we need to cast this term ID as a number here to ensure the `filteredByGroup` function always works:

```suggestion
			setCampaignGroup( parseInt( activeCampaignGroup ) );
```

I sometimes run into an issue where if I refresh the page or navigate to the Campaigns tab after setting an active group, the initial page load would show an empty list even if there were campaigns in the active group.

**Review by dkoo — CHANGES_REQUESTED** [review]
Ran into one bug when testing this step:

> 4. Refresh the page, observe the Campaign Groups UI defaults to the active group. Observe this is also true when switching from another tab back to Campaigns.

Details are in the comment below.

At a higher level, I wanted to ask what the thinking is behind only allowing a single active group? Consider the following scenario:

* I create two groups to organize my various campaigns: 1.) Newsletter, and 2.) Donation.
* It's pledge week, and I now want campaigns from both groups to be displayed.

Is the only way to accomplish the above to create a _third_ group called "Newsletter AND Donation" and apply it to all campaigns from both groups, then make that the active group? This seems counterintuitive to me and might force editors to create a lot of redundant groups, depending on how they organize their campaigns.

Suggestion: what if each group could be toggled active or inactive, with inactive being the default? Then all groups that are marked active can be shown on the front-end? 

I think this feature warrants a bit of extra discussion because of the fact that it's a breaking change and might force publishers to reconsider how they're organizing and using their campaigns.

**Comment by jeffersonrabb** [comment]
> Is the only way to accomplish the above to create a _third_ group called "Newsletter AND Donation" and apply it to all campaigns from both groups, then make that the active group? This seems counterintuitive to me and might force editors to create a lot of redundant groups, depending on how they organize their campaigns.

Campaigns can be added to multiple groups, which would be a much easier way to share a campaign across groups without creating new groups. 

> At a higher level, I wanted to ask what the thinking is behind only allowing a single active group? Consider the following scenario:
> 
> * I create two groups to organize my various campaigns: 1.) Newsletter, and 2.) Donation.
> * It's pledge week, and I now want campaigns from both groups to be displayed.
> 
> Is the only way to accomplish the above to create a _third_ group called "Newsletter AND Donation" and apply it to all campaigns from both groups, then make that the active group? This seems counterintuitive to me and might force editors to create a lot of redundant groups, depending on how they organize their campaigns.
> 
> Suggestion: what if each group could be toggled active or inactive, with inactive being the default? Then all groups that are marked active can be shown on the front-end?
> 
> I think this feature warrants a bit of extra discussion because of the fact that it's a breaking change and might force publishers to reconsider how they're organizing and using their campaigns.

I think you may be conflating expected use cases for _segments_ with _campaign groups_. "Newsletter" and "Donation" are more likely to be _segments_ than _campaign groups_. A campaign group is meant to represent the complete set of segments meant to be shown on a site at a given time, and is likely to contain a reasonably high number of campaigns (10?) each using a variety of segments. You might expect campaign groups to have names like "Production,"  "Experimental," maybe "Inactive." Overall I expect most sites will have very few campaign groups, perhaps only "active" and "inactive." Taking your example, you could have a Campaign Group called "Pledge Week," which is activated only for that specific period of time. Also, Campaigns can be members of multiple Campaign Groups, so rather than creating a Group called "Newsletter and Donation," you would just add the Campaigns to both groups, no need to create a hybrid group. Does that make sense?

File: assets/wizards/popups/views/popup-group/index.js (hunk 1 above)
Comment (jeffersonrabb) [reply assets/wizards/popups/views/popup-group/index.js]: I wasn't able to reproduce this on my local instance, but I _think_ I saw it on a production site I deployed this branch to. Here's the fix: https://github.com/Automattic/newspack-plugin/pull/775/commits/7b91839c9d8397081e1441c1996107bba6d22bf0

Comment (dkoo) [reply assets/wizards/popups/views/popup-group/index.js]: Thanks!

File: assets/wizards/popups/views/popup-group/index.js [hunk 2]
```diff
@@ -104,29 +126,98 @@ const PopupGroup = ( {
...
+						labelPosition="side"
+						disabled={ -1 === campaignGroups }
+					/>
+					{ campaignGroup > 0 && (
+						<SegmentationPreview
+							campaignGroups={ campaignGroup }
+							segment={ segmentId }
+							renderButton={ ( { showPreview } ) => (
+								<div className="newspack-campaigns__popup-group__filter-group-segmentation">
+									{ activeCampaignGroup !== campaignGroup && (
```
Comment (dkoo) [inline assets/wizards/popups/views/popup-group/index.js:159]:
```suggestion
									{ +activeCampaignGroup !== campaignGroup && (
```

File: assets/wizards/popups/views/popup-group/index.js [hunk 3]
```diff
@@ -104,29 +126,98 @@ const PopupGroup = ( {
...
+										<Button
+											isTertiary
+											isSmall
+											onClick={ () => manageCampaignGroup( campaignGroup ) }
+										>
+											{ __( 'Activate', 'newspack' ) }
+										</Button>
+									) }
+									{ campaignGroupExists && activeCampaignGroup === campaignGroup && (
```
Comment (dkoo) [inline assets/wizards/popups/views/popup-group/index.js:168]:
```suggestion
									{ campaignGroupExists && +activeCampaignGroup === campaignGroup && (
```

**Review by dkoo — CHANGES_REQUESTED** [review]
Thanks for explaining your vision of how groups should be used—definitely helps clarify things for me.

I see a couple of other places where the `activeCampaignGroup` var needs to be recast as well. Without these, in my site the "Activate" button never changes to "Deactivate".

File: assets/wizards/popups/views/popup-group/index.js (hunk 2 above)
Comment (jeffersonrabb) [reply assets/wizards/popups/views/popup-group/index.js]: I accidentally put this in a merge commit, but it's done in https://github.com/Automattic/newspack-plugin/pull/775/commits/99dd0a6a789c879b62a0e5a33643d1cda05aa0bb

File: assets/wizards/popups/views/popup-group/index.js (hunk 3 above)
Comment (jeffersonrabb) [reply assets/wizards/popups/views/popup-group/index.js]: Done in https://github.com/Automattic/newspack-plugin/pull/775/commits/99dd0a6a789c879b62a0e5a33643d1cda05aa0bb

**Comment by jeffersonrabb** [comment]
On hold pending completion of https://github.com/Automattic/newspack-plugin/pull/799

**Comment by dkoo** [comment]
Closing in favor of #804.

**Pull request closed without merging**